src/trace.zig

```diff
@@ -161,14 +171,40 @@ pub const Event = union(enum) {
 
     // Stack is a u32 since it must be losslessly encoded as a JSON integer.
     fn stack(event: *const Event) u32 {
-        const stack_base = event_stack_base.get(event.*);
-        const stack_offset: u32 = switch (event.*) {
-            .grid_read => |data| @intCast(data.iop),
-            .grid_write => |data| @intCast(data.iop),
-            else => 0,
-        };
-        assert(stack_offset < event_stack_cardinality.get(event.*));
-        return stack_base + stack_offset;
+        switch (event.*) {
+            .lookup_worker => |data| {
+                assert(data.index < event_stack_cardinality.get(event.*));
+                const stack_base = event_stack_base.get(event.*);
+                return stack_base + data.index;
+            },
+            .scan_tree => |data| {
+                assert(data.index < constants.lsm_scans_max);
+                // This event has "nested" sub-events, so its offset is calculated
+                // with padding to accommodate `scan_tree_level` events in between.
+                const stack_base = event_stack_base.get(event.*);
+                const scan_tree_offset = (constants.lsm_levels + 1) * data.index;
+                return stack_base + scan_tree_offset;
+            },
+            .scan_tree_level => |data| {
+                assert(data.index < constants.lsm_scans_max);
+                assert(data.level < constants.lsm_levels);
+                // This is a "nested" event, so its offset is calculated
+                // relative to the parent `scan_tree`'s offset.
+                const stack_base = event_stack_base.get(.scan_tree);
+                const scan_tree_offset = (constants.lsm_levels + 1) * data.index;
+                const scan_tree_level_offset = data.level + 1;
+                return stack_base + scan_tree_offset + scan_tree_level_offset;
+            },
+            inline .grid_read, .grid_write => |data| {
+                assert(data.iop < event_stack_cardinality.get(event.*));
+                const stack_base = event_stack_base.get(event.*);
+                return stack_base + @as(u32, @intCast(data.iop));
+            },
+            inline else => |data| {
+                comptime assert(@TypeOf(data) == void);
+                return event_stack_base.get(event.*);
+            },
```

Comment:
How about this:

```suggestion
            inline else => |data, event_tag| {
                comptime assert(@TypeOf(data) == void);
                return comptime event_stack_base.get(event_tag);
            },
```

Iiuc we use `inline else` rather than `else` so that we can assert `@TypeOf(data) == void`.

So since we are generating code for each variant anyway, may as well inline the stack base as a constant rather than needing to fetch it at runtime from the `EnumArray`.